panic while using --stdin-from-command

<!--

Welcome! - We kindly ask that you:

  1. Fill out the issue template below - not doing so needs a good reason.
  2. Use the forum if you have a question rather than a bug or feature request.

The forum is at: https://forum.restic.net

NOTE: Not filling out the issue template needs a good reason, as otherwise it
may take a lot longer to find the problem, not to mention it can take up a lot
more time which can otherwise be spent on development. Please also take the
time to help us debug the issue by collecting relevant information, even if
it doesn't seem to be relevant to you. Thanks!

The forum is a better place for questions about restic or general suggestions
and topics, e.g. usage or documentation questions! This issue tracker is mainly
for tracking bugs and feature requests directly relating to the development of
the software itself, rather than the project.

Thanks for understanding, and for contributing to the project!

-->


Output of `restic version`
--------------------------
restic 0.17.0 compiled with go1.22.5 on linux/amd64


What backend/service did you use to store the repository?
---------------------------------------------------------
local file system


Problem description / Steps to reproduce
----------------------------------------
restic init -r /tmp/test
restic backup -r /tmp/test --stdin-from-command
<!--
This section should include at least:

 * A description of the problem you are having with restic.

 * The complete command line and any environment variables you used to
   configure restic's backend access. Make sure to replace sensitive values!

 * The output of the commands, what restic prints gives may give us much
   information to diagnose the problem!

 * The more time you spend describing an easy way to reproduce the behavior (if
   this is possible), the easier it is for the project developers to fix it!
-->


Expected behavior
-----------------
no panic, a nice error message instead
<!--
Describe what you'd like restic to do differently.
-->

Actual behavior
---------------
enter password for repository: 
repository 5a50c088 opened (version 2, compression level auto)
created new cache in /home/xxx/.cache/restic
found 1 old cache directories in /home/xxx/.cache/restic, run `restic cache --cleanup` to remove them
no parent snapshot found, will read all files
[0:00]          0 index files loaded
panic: runtime error: slice bounds out of range [1:0]

goroutine 1 [running]:
github.com/restic/restic/internal/fs.NewCommandReader({0x1468450?, 0xc00003e550?}, {0xc0004020c0?, 0x0?, 0x0?}, {0x7b38a7fd9f38?, 0xc0004021e0?})
	/restic/internal/fs/fs_reader_command.go:33 +0x26a
main.runBackup({_, _}, {{{0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...}, ...)
	/restic/cmd/restic/cmd_backup.go:585 +0xa73
main.init.func3(0x1cac9e0, {0xc0004020c0, 0x0, 0x3})
	/restic/cmd/restic/cmd_backup.go:61 +0x116
github.com/spf13/cobra.(*Command).execute(0x1cac9e0, {0xc000402060, 0x3, 0x3})
	/home/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985 +0xaca
github.com/spf13/cobra.(*Command).ExecuteC(0x1cb2300)
	/home/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
	/home/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/home/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1034
main.main()
	/restic/cmd/restic/main.go:125 +0x265

<!--
In this section, please try to concentrate on observations, so only describe
what you observed directly.
-->

Do you have any idea what may have caused this?
-----------------------------------------------
I didn't specify any command after the --stdin-from-command flag
<!--
Did something noteworthy happen on your system, Internet connection, backend services, etc?
-->


Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
restic is an pretty stable piece of software - it's awesome
<!--
Answering this question is not required, but if you have anything positive to share, please do so here!
Sometimes we get tired of reading bug reports all day and a little positive end note does wonders.
Idea by Joey Hess, https://joeyh.name/blog/entry/two_holiday_stories/
-->

--- a/changelog/unreleased/issue-4975
+++ b/changelog/unreleased/issue-4975
@@ -1,0 +1,6 @@
+Bugfix: Prevent `backup --stdin-from-command` from panicking
+
+If --stdin-from-command is used, restic now checks whether there is a command behind it.
+
+https://github.com/restic/restic/issues/4975
+https://github.com/restic/restic/pull/4976

--- a/internal/fs/fs_reader_command.go
+++ b/internal/fs/fs_reader_command.go
@@ -29,6 +29,10 @@ type CommandReader struct {
 }
 
 func NewCommandReader(ctx context.Context, args []string, logOutput io.Writer) (*CommandReader, error) {
+	if len(args) == 0 {
+		return nil, fmt.Errorf("no command was specified as argument")
+	}
+
 	// Prepare command and stdout
 	command := exec.CommandContext(ctx, args[0], args[1:]...)
 	stdout, err := command.StdoutPipe()

--- a/internal/fs/fs_reader_command_test.go
+++ b/internal/fs/fs_reader_command_test.go
@@ -34,6 +34,11 @@ func TestCommandReaderInvalid(t *testing.T) {
 	test.Assert(t, err != nil, "missing error")
 }
 
+func TestCommandReaderEmptyArgs(t *testing.T) {
+	_, err := fs.NewCommandReader(context.TODO(), []string{}, io.Discard)
+	test.Assert(t, err != nil, "missing error")
+}
+
 func TestCommandReaderOutput(t *testing.T) {
 	reader, err := fs.NewCommandReader(context.TODO(), []string{"echo", "hello world"}, io.Discard)
 	test.OK(t, err)

